crates/storage/provider/src/transaction.rs

```diff
@@ -522,31 +492,25 @@ where
         if blocks.is_empty() {
             return Ok(())
         }
-        let tip = blocks.last().unwrap();
-        let new_tip_number = tip.number;
-        let new_tip_hash = tip.hash;
-        let expected_state_root = tip.state_root;
-
-        let fork_block_number = blocks.first().unwrap().number.saturating_sub(1);
+        let new_tip = blocks.last().unwrap();
+        let new_tip_number = new_tip.number;
 
-        let first_transition_id = self.get_block_transition(fork_block_number)?;
+        // Write state and changesets to the database
+        state.write_to_db(self.deref_mut())?;
 
-        let num_transitions = state.transitions_count();
+        let first_number = blocks.first().unwrap().number;
 
-        // Write state and changesets to the database
-        state.write_to_db(self.deref_mut(), first_transition_id)?;
+        let last = blocks.last().unwrap();
+        let last_number = last.number;
+        let last_hash = last.hash();
+        let last_state_root = last.state_root;
```

Comment:
nit: let's not call these "last" maybe? because the previous naming communicated that it is the "expected" one, which indeed happens to be the last one, but feels clearer to call it expected I think. no strong opinions

Reply:
renamed it to `expected_state_root`